internal/receiver/awscontainerinsightreceiver/k8sapiserver/k8sapiserver.go

```diff
@@ -0,0 +1,210 @@
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+)
+
+const (
+	lockName = "aoc-clusterleader"
+)
+
+type K8sAPIServer struct {
+	nodeName string //get the value from downward API
+	logger   *zap.Logger
+	cancel   context.CancelFunc
+	leading  bool
+}
+
+func New(logger *zap.Logger) *K8sAPIServer {
+	k := &K8sAPIServer{
+		logger: logger,
+	}
+
+	if err := k.start(); err != nil {
+		k.logger.Warn("Fail to start k8sapiserver", zap.Error(err))
+		return nil
+	}
+
+	return k
+}
+
+func (k *K8sAPIServer) GetMetrics() []pdata.Metrics {
+	var result []pdata.Metrics
+	if k.leading {
+		log.Printf("D! collect data from K8s API Server...")
+		timestamp := strconv.FormatInt(time.Now().UnixNano(), 10)
+		client := k8sclient.Get()
+
+		fields := map[string]interface{}{
+			"cluster_failed_node_count": client.Node.ClusterFailedNodeCount(),
+			"cluster_node_count":        client.Node.ClusterNodeCount(),
+		}
+		tags := map[string]string{
+			common.MetricType: common.TypeCluster,
+			common.Timestamp:  timestamp,
+			common.Version:    "0",
+		}
+		if k.nodeName != "" {
+			tags["NodeName"] = k.nodeName
+		}
+		md := common.ConvertToOTLPMetrics(fields, tags, k.logger)
+		result = append(result, md)
+
+		for service, podNum := range client.Ep.ServiceToPodNum() {
+			fields := map[string]interface{}{
+				"service_number_of_running_pods": podNum,
+			}
+			tags := map[string]string{
+				common.MetricType:   common.TypeClusterService,
+				common.Timestamp:    timestamp,
+				common.TypeService:  service.ServiceName,
+				common.K8sNamespace: service.Namespace,
+				common.Version:      "0",
+			}
+			if k.nodeName != "" {
+				tags["NodeName"] = k.nodeName
+			}
+			md := common.ConvertToOTLPMetrics(fields, tags, k.logger)
+			result = append(result, md)
+		}
+
+		for namespace, podNum := range client.Pod.NamespaceToRunningPodNum() {
+			fields := map[string]interface{}{
+				"namespace_number_of_running_pods": podNum,
+			}
+			tags := map[string]string{
+				common.MetricType:   common.TypeClusterNamespace,
+				common.Timestamp:    timestamp,
+				common.K8sNamespace: namespace,
+				common.Version:      "0",
+			}
+			if k.nodeName != "" {
+				tags["NodeName"] = k.nodeName
+			}
+			md := common.ConvertToOTLPMetrics(fields, tags, k.logger)
+			result = append(result, md)
+		}
+	}
+	return result
+}
+
+func (k *K8sAPIServer) start() error {
+	var ctx context.Context
+	ctx, k.cancel = context.WithCancel(context.Background())
+
+	k.nodeName = os.Getenv("HOST_NAME")
+	if k.nodeName == "" {
+		return errors.New("Missing environment variable HOST_NAME. Please check your YAML config.")
```

Comment:
the golang error msg starts with lowercase. :)

Reply:
`HOST_NAME` is actually defined in the yaml file for cloudwatch agent

https://github.com/aws-samples/amazon-cloudwatch-container-insights/blob/master/k8s-deployment-manifest-templates/deployment-mode/daemonset/container-insights-monitoring/cwagent/cwagent-daemonset.yaml#L36-L39